inbm/dispatcher-agent/dispatcher/sota/cancel.py

```diff
@@ -0,0 +1,86 @@
+"""
+    Method to handle cancel request
+
+    Copyright (C) 2017-2024 Intel Corporation
+    SPDX-License-Identifier: Apache-2.0
+"""
+
+import logging
+from typing import Optional
+from threading import Event
+from inbm_lib.xmlhandler import XmlHandler
+from threading import Thread
+from .constants import SOTA_CACHE
+from ..constants import OtaType
+from dispatcher.dispatcher_exception import DispatcherException
+from dispatcher.packagemanager.local_repo import DirectoryRepo
+from dispatcher.dispatcher_broker import DispatcherBroker
+from dispatcher.common.result_constants import Result, CODE_OK, CODE_BAD_REQUEST
+
+logger = logging.getLogger(__name__)
+
+
+def cancel_thread(type_of_manifest: str, parsed_head: XmlHandler, thread_list: list[Thread],
+                  type_of_active_manifest: Optional[str], active_thread_parsed_head: Optional[XmlHandler],
+                  dispatcher_broker: DispatcherBroker, cancel_event: Event) -> bool:
+    """
+    Cancel the current active thread by sending the terminate signal.
+
+    @param type_of_manifest: type of the request
+    @param parsed_head: The root parsed xml
+    @param thread_list: List of the active thread
+    @param type_of_active_manifest: type of the request on running thread
+    @param active_thread_parsed_head: The root parsed xml of running thread
+    @param dispatcher_broker: DispatcherBroker object used to communicate with other INBM services
+    @param cancel_event: Event used to stop the downloading process
+    @return: True if the request has been processed; False if no request has been handled.
+    """
+    if type_of_manifest == 'ota':
+        header = parsed_head.get_children('ota/header')
+        ota_type = header['type']
+        resource = parsed_head.get_children(f'ota/type/{ota_type}')
+        if ota_type == OtaType.SOTA.name.lower():
+            sota_mode = resource.get('mode', None)
+            if sota_mode == 'cancel':
+                logger.debug(f"Receive sota cancel request.")
+                # If the active thread is not SOTA download-only, forbid the cancel request.
+                if type_of_active_manifest and active_thread_parsed_head:
+                    if not is_active_ota_sota_download_only(type_of_active_manifest, active_thread_parsed_head):
+                        dispatcher_broker.send_result(
+                            str(Result(CODE_BAD_REQUEST, "Current thread is not SOTA download-only. "
+                                                         "Cannot proceed with the cancel request.")))
+                        return True
+                else:
+                    dispatcher_broker.send_result(str(Result(CODE_BAD_REQUEST, "Running thread manifest not found.")))
+                    return True
+
+                # The list should only contain one OTA process.
+                for thread in thread_list:
+                    if thread.is_alive():
+                        cancel_event.set()
```

Comment:
One potential issue here: `cancel_event` will remain set after this. It needs to be reset once it's been handled by the thread.

A way you could cleanly work around this is by storing one event per thread, maybe using `_thread_list`? Then you wouldn't need to reset at all--you'd have a fresh event for each thread.

Reply:
I used to have a line to reset the flag after thread.join. Somehow it's missing here. 
How about reset the cancel_event after the OTA thread.join? 

                for thread in thread_list:
                    if thread.is_alive():
                        cancel_event.set()
                        # Wait thread to gracefully exit
                        logger.debug(f"Waiting thread to exit...")
                        thread.join()
                        logger.debug(f"Request cancel complete.")
                        # Reset the event flag 
                        cancel_event.clear()
                return True

Reply:
> I used to have a line to reset the flag after thread.join. Somehow it's missing here. How about reset the cancel_event after the OTA thread.join?

I'm now thinking that if each thread has its own event, this won't be a problem.

Reply:
@gblewis1  I have updated the method to only cancel the request in the middle of downloading process. For assigning a unique event flag to each thread, I'm thinking to have a class below to manage the thread and set the event flag as necessary. I realize that this modification will require the event being passed through many methods, such as do_install -> _do_ota_update, _validate_pota_manifest, invoke_sota. This will require time to validate the changes and to update the unit tests accordingly. Could we create a story for this enhancement to address it in the future? 

```
class ThreadInstance:
    """ThreadInstance class manages the lifecycle of a thread for a given task.
    @param target: Method to be executed
    @param args: Arguments to be passed to the target function
    """
    def __init__(self, target: Callable[..., object], args: Iterable[Any]) -> None:
        self.cancel_event = Event()
        self.thread = Thread(target=target, args=(args, self.cancel_event))

    def set_event(self) -> None:
        self.cancel_event.set()

    def is_alive(self) -> bool:
        return self.thread.is_alive()

    def join(self) -> None:
        self.thread.join()

    def start(self) -> None:
        self.thread.setDaemon(True)
        self.thread.start()
```

Reply:
> @gblewis1 I have updated the method to only cancel the request in the middle of downloading process. For assigning a unique event flag to each thread, I'm thinking to have a class below to manage the thread and set the event flag as necessary. I realize that this modification will require the event being passed through many methods, such as do_install -> _do_ota_update, _validate_pota_manifest, invoke_sota. This will require time to validate the changes and to update the unit tests accordingly. Could we create a story for this enhancement to address it in the future?


As long as the download cancel is working reliably in our new use case, I think it's fine to create a story to enhance it later, as the new use case + testing is top priority right now.